Puzzles System - first cut

Henddher opened the pull request:
- `PuzzlePartObject`: typeclass for puzzle parts and results.
- `PuzzleRecipeObject`: typeclass to store prototypes of parts and results.
- `PuzzleSystemCmdSet`: commands to create, arm and resolve puzzles.

Puzzles System - Provides a typeclass and commands for
objects that can be combined (i.e. 'use'd) to produce
new objects.

A Puzzle is a recipe of what objects (aka parts) must
be combined by a player so a new set of objects
(aka results) are automatically created.

Consider this simple Puzzle:

    orange, mango, yogurt, blender = fruit smoothie

As a Builder:

    @create/drop orange
    @create/drop mango
    @create/drop yogurt
    @create/drop blender

    @puzzle smoothie puzzle, orange, mango, yogurt, blender = fruit smoothie
    ...
    Puzzle smoothie puzzle (#1234) created successfuly.

    @destroy/force orange, mango, yogurt, blender

    @armpuzzle #1234
    Part orange is spawned at ...
    Part mango is spawned at ...
    ....
    Puzzle smoothie puzzle (#1234) has been armed successfully

As Player:

    use orange, mango, yogurt, blender
    ...
    Genius, you blended all fruits to create a yummy smoothie!

Details:

Puzzles are created from existing objects. The given
objects are introspected to create prototypes for the
puzzle parts. These prototypes become the puzzle recipe.
(See PuzzleRecipeObject and @puzzle command).

At a later time, a Builder or a Script can arm the puzzle
and spawn all puzzle parts (PuzzlePartObject) in their
respective locations (See @armpuzzle).

A regular player can collect the puzzle parts and combine
them (See use command). If player has specified
all pieces, the puzzle is considered solved and all
its puzzle parts are destroyed while the puzzle results
are spawened on their corresponding location.

Henddher commented:
Hi @Griatch and @v-legoff @vlegoff 

I just pushed a significant chunk of changes for the puzzle system 🎉.

I went with a simplified version of all your suggestion. I implemented the concept of **recipes** by adding a new`PuzzleRecipeObject` typeclass. I am using **tags** to group all `recipes` and `parts` so they can be found and quickly filtered. Finally, I added a command set to create recipes, spawn puzzle parts from their puzzle's recipe and finally spawn/destroy objects when a puzzle is resolved.

Griatch commented on evennia/contrib/puzzles.py:
```diff
@@ -0,0 +1,532 @@
+
+    orange, mango, yogurt, blender = fruit smoothie
+
+As a Builder:
```
I find this way of creating objects to build the recipe only to then delete them to be pretty cumbersome. I discuss the tag-alternative in answer to your question below.

Henddher replied:
Not quite ... see below

Griatch commented on evennia/contrib/puzzles.py:
```diff
@@ -0,0 +1,472 @@
+            )
+
+        # Find all puzzles by puzzle name
+        # FIXME: we rely on obj.db.puzzle_name which is visible and may be cnaged afterwards. Can we lock it and hide it?
```
@Henddher As mentioned, if you use tag-names in your recipe (instead of, as now, full objects), you could use the tag's category to identify it as a component of a particular puzzle or just as being a generic puzzle piece at all. As long as the result of the puzzle is also tagged as a puzzle-piece you could then use that in another puzzle as much as you wanted. 

So the creation of a recipie would then be a lot easier (example of syntax)

```
@puzzle smoothie: orange + mango + yoghurt + blender = fruit smoothie;smoothie:typeclasses.objects.Smoothie
```
This creates a new Recipie (a Script) named `smoothie` holding the recipie tag-names) and the instructions for how to build the resulting object. You could now use `@armrecipe` to build these components or manually build them and tag each with a Tag `orange` and category `recipe_ingredient` or something. Then the `orange` component could actually be named (have a `key`) `An yellow-red fruit` or `A rotting orange` or "Peeled orange" to make the puzzle harder or to match your game world. It also means that multiple objects could act as components in the recipe if you wanted to, as long as they have the right tag. "A can of preserved orange slices" works just as well as "A big orange" as long as they are both tagged right. 

All you now need to do is make sure the resulting created object is created with a matching tag. Maybe you'll default to doing so (so give the `fruit smoothie` a Tag `fruit smoothie` always). You can then naturally use this with another recipe: 

```
@recipe fruit_grenade:fruit smoothie + exploding bananas + metal can = fruit grenade:typeclasses.objects.FruitGrenade
```

Since all you specify for your recipe are strings, it becomes very easy to set up and your entire code becomes a lot simpler. Of course, here I suggested specifying typeclass for the resulting object since I expect it to be created on the fly when the recipe is completed. You could also give a prototype if you preferred.

Henddher commented on evennia/contrib/puzzles.py:
```diff
@@ -0,0 +1,533 @@
+from evennia.commands.default.muxcommand import MuxCommand
+from evennia.utils.utils import inherits_from
+from evennia.utils import search, utils, logger
+from evennia.utils.spawner import spawn
```
I'm adding tests as we speak so that should mitigate regressions when going to 0.8.

Henddher commented:
Thank you @Griatch!!

https://github.com/evennia/evennia/pull/1653#discussion_r214806663
> I find this way of creating objects to build the recipe only to then delete them to be pretty cumbersome. I discuss the tag-alternative in answer to your question below.

I see your point.  However, IMHO, I think it would be actually better to use any object that already exists in the world - disregarding when or how or whom created it.

In the first PR #589, you pointed out the advantage of puzzle parts not knowing anything about puzzles (https://github.com/evennia/evennia/pull/589#issuecomment-57127115). 

> The idea here is that the recipe is defined outside the objects themselves, for example in a module that users can load (see for example the way most of Evennia's gamesrc/conf modules work, or more precisely how src.utils.spawner works). With a recipe, each object does not need to be aware that they are "combinable". 

https://github.com/evennia/evennia/pull/1653#discussion_r214810110
The new way you suggested (i.e. sword example with different blades) , albeit being very flexible, it adds an extra burden to the Builder: housekeeping of tags in objects as he/she needs to *tag* world objects so they can participate in puzzles.

### Let me explain why I think using existing objects would be better:

As a Builder, one creates objects with full descriptions, locks, locations, etc. i.e. one builds normal *world* objects. 

As a Builder (of puzzles), I would like to think that a puzzle-recipe is just another plain old world object.

Under this paradigm, the Builder would only need to create puzzle recipes by **using** existing world objects without having to make any changes to those objects nor having to create new ones either. IOW, no extra housekeeping beyond *vanilla* object creation.

In fact, the original objects would **not** need to be altered in any way because the puzzle-recipe would *remember* all these objects prototypes precisely so when the puzzle gets armed, *clones* of all those objects will appear in the world - fully described, with locks, etc. etc.

One obvious advantage is that the `@puzzle` command would stay as simple as it is currently. Its sole purpose would be to *record* the full prototypes of its declared parts and results. The command would have *nothing* to do with what the objects look and behave like - *at all*. 

From the players perspective, there are no drawbacks:

Any player could trigger the automatic arming of a puzzle, the parts would be spawned fully described, with locks, location, contents, etc. etc. exactly wherever they are supposed to be for that specific puzzle. That player, or **any** other player would be able to collect the pieces and resolve the puzzle. Multiple identical puzzles could be active at the same time triggered by multiple players. The parts would still be used across puzzles. Whenever a puzzle got resolved, its resulting parts would be spawned precisely how they are supposed to be.

In summary, Builder would only need to use vanilla builder commands to create puzzles from new or existing objects and there would be no extra maintenance of tags related to puzzle parts and puzzle results.

WDYT?

Griatch commented:
@Henddher I see what you mean, concerning the design of the puzzle. There are a few aspects of it which concerns me however: 
1) Specificity. With this set-up, it's correct that the objects don't need to know that they are part of the puzzle. It however also ties each puzzle-component piece to a very specific, individual object instance - a clone as you put it. The difference with a tagging solution is that you can ascribe _classes_ of objects that all work for the same puzzle - the tagging makes sure to make the puzzle-part's identifier unique in the eyes of the puzzle *without* requiring the objects to be identical. This opens up for flexibility in the puzzle creation - you can use both 'can of olives' and 'single olive' to make a drink if you want, assuming these entities are tagged right. 
2) Debugging and maintenance. Let's say you have multiple olive objects in the world. Or an olive gets deleted by accident. Or some other bug appears. In this variation only olives created specifically for this recipe are useful. In the other alternative, only objects with the right tag are useful. I'd argue that it's easier to detect a lack of tag than to go back to the recipe to find what specific prototype instance it requires.
3) Lack of extendability. I know this is not your goal with this PR, but a flexible-enough recipe system could just as well work for a crafting system as for a puzzle system. That other use is impossible if each recipe is tied to a specific object rather than to a class of objects. 

But that said, it's your contrib and you have a very specific use-case in mind - and for that the current solution does work. I won't force you into something beyond the scope you planned. So if you prefer the object-based design that's fine. :) I'd need to review the code again though after your latest updates.

Henddher commented on evennia/contrib/puzzles.py:
```diff
@@ -0,0 +1,533 @@
+        self.db.puzzle_name = puzzle_name
+
+
+class PuzzleRecipeObject(DefaultObject):
```
Hi @Griatch, I tried replacing `DefaultObject` with `DefaultScript` but got stuck because `evennia.utils.search.search_script()` can only search by `obj.key` and by `obj.dbref`. 

In this case, I need to search by `attribute_name`.

I'm not familiar with Django query syntax but I certainly wouldn't mind "adding" such functionality to a new helper function in `utils.search` *except* for the 0.8 upcoming release.

Being unable to search by `attribute_name` is a roadblock for me.

Suggestions?

Griatch replied:
@Henddher This functionality is available for all typeclassed entities through a manager method, although not through the `search_script` wrapper. Use it like this: 

```python
from evennia import ScriptDB
scripts = ScriptDB.objects.get_by_attribute(key="foo", category="bar", value=None)
```
where `key`, `category` and `value` are all optional and are ways to find scripts only with that Attribute or attribute-value attached to them. The definition is here: https://github.com/evennia/evennia/blob/master/evennia/typeclasses/managers.py#L96

Henddher replied:
Thanks @Griatch - that worked. 

I `git grep`ed and found a thin wrapper of it in `utils.search`. 

https://github.com/evennia/evennia/blob/a53957096f33ce991a52921fe9a26e1984367ede/evennia/utils/search.py#L208-L209

Perhaps that one's "better" for abstraction purposes?

Henddher commented:
Thank you @Griatch

I understand your concerns https://github.com/evennia/evennia/pull/1653#issuecomment-418616304.

With regards to:
> .. It however also ties each puzzle-component piece to a very specific, individual object instance ...

That's indeed a handicap of my current implementation. Specifically, when a recipe is armed and its parts are spawn, the assignment `self.db.puzzle_name = puzzle_name` occurs for each puzzle part and result (see line 127: `mark_as_puzzle_member` function). 

It is a simple attempt to *optimize* the matching algorithm: a part belongs to a puzzle if `part.db.puzzle_name == recipe.db.puzzle_name`.

For the crafting system you envisioned, Builder would need to add tags to new objects and older objects if he/she intends that such objects are to be "compatible" with new puzzles. 

I think the fundamental problem is modeling a Many-to-Many relationship between puzzle recipes, parts and results vs Easy maintenance. 

Tags do solve the many-to-many problem because a puzzle recipe and its parts (and results) can share the same tag. The only questionable aspect is the *extra burden* of Builders having to add Tags to old and new objects as new puzzle recipes are added.

### I'd like to get these issues ironed out before the official code-review:

- [ ] Remove the `one-to-one` part vs puzzle issue inherent to `part.db.puzzle_name == puzzle_recipe.db.puzzle_name`. 

- [ ] Enhance the prototype generation and recipe & parts matching so that Rooms, Exits, and even NPCs can be part of a puzzle ... although I am not sure if Locks would provide the same functionality/behavior out of the box.

### Still WIP :)

Griatch commented:
@Henddher I won't rehash this too much, but as for the "extra burden" of tagging: You could wrap the tagging aspect completely in your puzzle-command structure if you wanted to. 

For example you could have a command  `puzzle/add puzzlename=<obj>` that just tags an existing object correctly with tag/category. 

For easy maintenance you could then let `puzzle puzzlename` list info about the puzzle, including all existing objects with that tag - you could even let the same command error-check and report if one of the components does not exist. A more general solution would be to show how many many complete puzzle-results could be created from the existing components in the world (and if there are any lack of a given component to make a 'complete' solution). 

Add to that some of your existing helper commands to auto-create all components of a recipie and I don't see why this would be any 'extra burden' on the builder - they don't need to know anything about tagging at all, it'd be just an under-the-hood implementation detail ...

Griatch commented on evennia/commands/default/muxcommand.py:
```diff
@@ -118,6 +118,11 @@ def parse(self):
             lhs, rhs = [arg.strip() for arg in args.split('=', 1)]
             lhslist = [arg.strip() for arg in lhs.split(',')]
             rhslist = [arg.strip() for arg in rhs.split(',')]
+        # eliminate all empty-strings
```
It looks strange but if you think about what rhslist/lhslist do, is that they split the contents on the left-hand- and right-hand-side of the `=` by commas. What you have entered there is two empty strings separated by a comma. So while it looks weird that is correct syntax; it allows for example for entering empty fields in multi-comma input arrays.

Henddher replied:
That's a good point.
I'll revert those changes and remove my commented out test case.

Henddher commented:
> You could wrap the tagging aspect completely in your puzzle-command structure if you wanted to.
> For example you could have a command `puzzle/add puzzlename=<obj>` that just tags an existing object correctly with tag/category.

That's true! 
Builders don't need to `@tag` by hand when we can give them helper commands.

> For easy maintenance you could then let puzzle puzzlename list info about the puzzle, including all existing objects with that tag - you could even let the same command error-check and report if one of the components does not exist. A more general solution would be to show how many many complete puzzle-results could be created from the existing components in the world (and if there are any lack of a given component to make a 'complete' solution).

I created 2 helper commands (`@lspuzzlerecipes` and `@lsarmedpuzzles` [1]) but I could augment their scope as you suggested.

@Griatch, I will give that a shot!

I reached [97% coverage](https://coveralls.io/builds/19013863/source?filename=/home/travis/build/evennia/evennia/evennia/contrib/puzzles.py) so I can evaluate alternative implementations with *low* risk 😄 

Thanks!

[1] Here is the output without colors (coming from my son's puzzles :P ):

```
------------------------------------------------------------
Puzzle name: Happygrill
             Happy Burger(#371) at Jp's super secret secret room of secrets(#228)
------------------------------------------------------------
Puzzle name: Sacrifice
      Poop Emoji Keychain(#550) at                The Figure(#432)
------------------------------------------------------------
Puzzle name: Happy Grill Puzzle
               Dead Grill(#349) at Jp's super secret secret room of secrets(#228)
             Happy Burger(#350) at Jp's super secret secret room of secrets(#228)
                 Recorder(#351) at Jp's super secret secret room of secrets(#228)
------------------------------------------------------------
Puzzle name: Spoonstore
             Garfield DVD(#501) at           The Secret Hall(#508)
------------------------------------------------------------
Puzzle name: spark
                    spark(#603) at                    garage(#136)
                    stone(#604) at                    garage(#136)
                    flint(#605) at                    garage(#136)
------------------------------------------------------------
Puzzle name: Happygrillpuzzle
                 Recorder(#570) at     The Jungle of Passion(#323)
------------------------------------------------------------
------------------------------------------------------------
Puzzle 'spark' spark(#1)
Parts
 -      typeclass: evennia.contrib.puzzles.PuzzlePartObject
             tags: [('puzzle_member', 'puzzles')]
         location: garage
              key: stone
             desc: You see nothing special.
 -      typeclass: evennia.contrib.puzzles.PuzzlePartObject
             tags: [('puzzle_member', 'puzzles')]
         location: garage
              key: flint
             desc: You see nothing special.
Results
 -      typeclass: evennia.contrib.puzzles.PuzzlePartObject
             tags: [('puzzle_member', 'puzzles')]
         location: garage
              key: spark
             desc: You see nothing special.
 -      typeclass: evennia.contrib.puzzles.PuzzlePartObject
             tags: [('puzzle_member', 'puzzles')]
         location: garage
              key: stone
             desc: You see nothing special.
 -      typeclass: evennia.contrib.puzzles.PuzzlePartObject
             tags: [('puzzle_member', 'puzzles')]
         location: garage
              key: flint
             desc: You see nothing special.
------------------------------------------------------------
------------------------------------------------------------
Puzzle name: Happygrill
             Happy Burger(#371) at Jp's super secret secret room of secrets(#228)
------------------------------------------------------------
Puzzle name: Sacrifice
      Poop Emoji Keychain(#550) at                The Figure(#432)
------------------------------------------------------------
Puzzle name: Happy Grill Puzzle
               Dead Grill(#349) at Jp's super secret secret room of secrets(#228)
             Happy Burger(#350) at Jp's super secret secret room of secrets(#228)
                 Recorder(#351) at Jp's super secret secret room of secrets(#228)
------------------------------------------------------------
Puzzle name: Spoonstore
             Garfield DVD(#501) at           The Secret Hall(#508)
------------------------------------------------------------
Puzzle name: spark
                    spark(#603) at                    garage(#136)
                    stone(#604) at                    garage(#136)
                    flint(#605) at                    garage(#136)
------------------------------------------------------------
Puzzle name: Happygrillpuzzle
                 Recorder(#570) at     The Jungle of Passion(#323)
------------------------------------------------------------
```

samjaninf commented:
This would make a sick crafting system. Crafting items/weapons/potions. Hmmm, with a little modification it could even be a little more in-depth, such as denoting the prime piece of the recipe, without that it doesn't work at all, but if they only have atleast say 50% of all the pieces then it would still work but at reduced level. Maybe even be able to add more than the base amount and achieve some sort of bosted item/recipe.

All in good time. Great idea on the Puzzle components @Henddher and thanks for all the hard work!

Griatch commented:
@Henddher I see you making changes, but I don't know status since this is a WIP thing. You'll need to let me know when you feel ready for another review, just so you know.

Henddher commented:
Hi @Griatch,

I'm still plowing through the list of "features" I'd like to incorporate .... so still WIP 🚜 🏭 

I was able to remove the single part-to-puzzle dependency when I moved to tags. I still have `part.db.puzzle_name` and `result.db.puzzle_name` but they aren't truly used. 

With respect to your https://github.com/evennia/evennia/pull/1653#issuecomment-421247718
> For example you could have a command puzzle/add puzzlename=<obj> that just tags an existing object correctly with tag/category.

At first, I ran into an issue to accomplish that.  In my current implementation, each puzzle part and puzzle result is expected to be typeclass `evennia.contrib.puzzles.PuzzlePartObject` so tagging an existing object (not inherited from it) would fail to incorporate that object into an existing recipe. The class is checked when the puzzle is being solved (`use` command). On top of that, I checked if an object could have multiple `typeclass_path` but such doesn't seem to be possible.

My alternative approach has been to modify the existing recipe in place. Basically, I take the existing object, derive its prototype and add it to the list of part or results:

`@puzzleedit/addpart ...` and `@puzzleedit/addresult ...`.

It all works well.

At this point, I think I can throw away `PuzzlePartObject` and ignore typeclass in all puzzle operations so your recommendation will be met.

I will need to modify both `@puzzleedit/addpart` and `@puzzleedit/addresult` so while the recipe is being modified, the passed in object will also be tagged and become part of existing armed puzzles.

So far, it's all coming together nicely I think 😄

Henddher commented on evennia/contrib/puzzles.py:
```diff
@@ -0,0 +1,747 @@
+    """
+    protodef = {
+        # FIXME: Don't we need to honor ALL properties? attributes, contents, etc.
+        'prototype_key': '%s(%s)' % (obj.key, obj.dbref),
```
Added `prototype_key` but tests still fail and it seems to be related to `tags`. 

See line 114 below and how I used to deal with `tags` in 0.7.0. Perhaps there is something else I need to do in 0.8.0?

Henddher commented:
Hi @Griatch 

I just pull from master (0.8.0) and encountered problems with the new prototypes framework (as you first mentioned https://github.com/evennia/evennia/pull/1653#discussion_r214750225).

Basically, I modified my `proto_def` function and added `prototype_key` but it seems like there is still some issues with how object tags are stored. It all worked well in 0.7.0.

See https://github.com/evennia/evennia/pull/1653/files/84004b7d3721024720ea5766ef4090a97f23e553

Henddher commented:
After some debugging, I came to this:

![screen shot 2018-10-06 at 9 07 35 am](https://user-images.githubusercontent.com/3991010/46572209-663dfe80-c947-11e8-8a68-ea481fd40ad4.png)

Is line 714 correct?
I think not, `tags` is `[]` from line 713. Shouldn't it be `for (...) in val:`

https://github.com/evennia/evennia/blob/41aca0c73969442412531308620c23261aa4c929/evennia/prototypes/spawner.py#L712-L715

Henddher commented:
>  `tags` require a full tuple always. That said, for those entering prototypes manually I can see the point of converting the older, looser format to the strict one (setting e.g. `(tagname, None, None)` when the tuple is not given) as a way to make it more convenient to manually create prototypes.
>
> I created a feature request for that conversion here: #1675 But when you enter prototypes codedly like here, you are best off to use the real format.

However, I still think that the loop I mentioned is not quite correct:

https://github.com/evennia/evennia/blob/41aca0c73969442412531308620c23261aa4c929/evennia/prototypes/spawner.py#L712-L715

I compared the output of `spawner.prototype_from_object(obj)` and the output of `puzzles.proto_def(obj)` and `spawner.prototype_from_object(obj)` doesn't seem to be honoring the tags.  See this `ipdb` screenshot:

![screen shot 2018-10-06 at 12 20 34 pm](https://user-images.githubusercontent.com/3991010/46574052-559b8180-c963-11e8-9fbe-09cde6d2e24f.png)

I will try to fix `spawner.prototype_from_object(obj)` and let you know if I ran into other issues. I believe that `spawner.prototype_from_object(obj)` should replace `puzzles.proto_def(obj)` and it should all work correctly. No?

Griatch commented:
@Henddher I can't reproduce your second issue: As a test I assigned two tags on myself: 

    @tag me = foo
    @tag me = bar 

Then I tried to get them using your two methods: 

    !self.tags.get(return_tagobj=True, return_list=True)
    >>> self.tags.get(return_tagobj=True, return_list=True)
    [<Tag: <Tag: foo>>, <Tag: <Tag: bar>>]
    
    !self.tags.all()
    >>> self.tags.all()
    ['foo', 'bar']

Which seems to be as expected.

Henddher commented:
> Which seems to be as expected.

Agree but not what I am getting on my end. 

Perhaps I'm to pull from master again?

```
commit 84004b7d3721024720ea5766ef4090a97f23e553 (HEAD -> puzzles, origin/puzzles)
Author: Henddher Pedroza <henddher.pedroza@gmail.com>
Date:   Sat Oct 6 08:00:23 2018 -0500

    0.8.0 port

commit 76771a9f30b6e6e881a34106f4e075a195e417ff
Merge: 7365f56d 41aca0c7
Author: Henddher Pedroza <henddher.pedroza@gmail.com>
Date:   Sat Oct 6 06:56:50 2018 -0500

    Merge branch 'master' of https://github.com/evennia/evennia into puzzles

commit 7365f56db89789fa451bfdf7125f1b4b4fa9096b
Merge: e92e830f 8c0dca25
Author: Henddher Pedroza <henddher.pedroza@gmail.com>
Date:   Sat Oct 6 06:56:24 2018 -0500

    Merge conflict

commit 41aca0c73969442412531308620c23261aa4c929 (remote/master)
Merge: 8c0dca25 0fd6b1be
Author: Griatch <griatch@gmail.com>
Date:   Sat Oct 6 05:09:19 2018 +0200

    Merge pull request #1672 from friarzen/master-historyjs-fix
    
    Fix #1668 - up arrow key regression
```

Henddher commented:
@Griatch, the following cannot be correct. The loop will never execute and `tags` will always be `[]`.
 
https://github.com/evennia/evennia/blob/41aca0c73969442412531308620c23261aa4c929/evennia/prototypes/spawner.py#L712-L715

Henddher commented:
I just stepped through `spawner.spawn()` and `batch_create_object(*objsparams)` returned the new object with none of the passed in tags.

https://github.com/evennia/evennia/blob/cb8441bf3a9ed515fad0d76052e6d13503dd79bc/evennia/prototypes/spawner.py#L610-L622

I am suspecting that the issue is due to 2 tags with same value but one with a `None` category and the other with a string category.

![screen shot 2018-10-06 at 6 37 48 pm](https://user-images.githubusercontent.com/3991010/46576727-f8b9be80-c996-11e8-84a3-4dd9aa496a96.png)

Henddher commented:
Hi @Griatch,

I'm still trying to figure out what's happening with the tags. It seems like they aren't being dropped as I first thought.  However, `.tags.all()` and `.tags.get(return_list=True, return_tagobj=True)` are in fact producing **different outputs**.

I'm `ipdb`'ing here:

https://github.com/evennia/evennia/blob/cb8441bf3a9ed515fad0d76052e6d13503dd79bc/evennia/prototypes/spawner.py#L608-L623

And this is the output:

![screen shot 2018-10-07 at 10 42 41 am](https://user-images.githubusercontent.com/3991010/46583511-2dbd2400-ca1e-11e8-923b-599a96198b54.png)

And of course, I'm patching the questionable loop I mentioned earlier https://github.com/evennia/evennia/pull/1653#issuecomment-427606975

```diff
diff --git a/evennia/prototypes/spawner.py b/evennia/prototypes/spawner.py
index ce089441..fd9b4545 100644
--- a/evennia/prototypes/spawner.py
+++ b/evennia/prototypes/spawner.py
@@ -714,8 +714,8 @@ def spawn(*prototypes, **kwargs):
 
         val = prot.pop("tags", [])
         tags = []
-        for (tag, category, data) in tags:
-            tags.append((init_spawn_value(val, str), category, data))
+        for (tag, category, data) in val:
+            tags.append((init_spawn_value(tag, str), category, data))
 
         prototype_key = prototype.get('prototype_key', None)
         if prototype_key:
```

Griatch commented:
@Henddher I've looked at this but not pushed any fix yet. The reason the tags.all() and tags.get() are outputting different outputs is because the latter will interpret not setting the `category` as the `category` being `None`. So you are right, that the category was the culprit there. I've expanded the `tags.all()` call with a `return_objs` keyword instead, which makes it a lot easier to get whatever combination of key/category/data without needing a lot of if statements. 

I'm currently bogged down in another bug but will try to push a fix for this soon.

Henddher commented:
Awesome @Griatch !

I've got good news too: 

I took care of the `tags`/`val` processing loop and together with removing `'prototype_key'` from my puzzle proto_def `dict`s, the puzzles module is back in business (100% tests passing).

I've pushed my changes. 

Please note that my *fix* to the `tags`/`val` processing loop doesn't handle every kind of tag tuple. IOW, it only expects tag tuples having `tag` (`str`), `category` and `data`. Any other combination of the above will fail. So if someone passed `tags` output from `.tags.all(return_key_and_category=True)` to `spawn()`, the object creation passed down to `batch_create_object` would not work as expected.

Griatch commented:
@Henddher I have pushed the fixes to the tag handler and spawn now; I did not use your changes to core (if you want to contribute such changes, they should have gone in a separate PR - mixing contrib code with core changes is not a good idea), so you should be able to remove them from your code. I tested with your branch and unit tests passed but since you have mixed in core changes it confuses the issue somewhat.

Henddher commented:
> ...hey should have gone in a separate PR - mixing contrib code with core changes is not a good idea)

Good point. 

I pulled from master and my fix is now replaced with yours from master branch. It's all good!

Griatch commented:
@Henddher Thanks for the help tracking this down, hopefully works better now :)

Henddher commented:
I think I'm almost done with this contribution 🎉 

From my original list of TODOs, there are only 2 *major* points I'd like to tackle (and 1 minor):

- [ ] 1. allow parts and results to be exits, rooms and even NPCs.
- [ ] 2. honoring parts' and/or results' contents.

1. is somehow natural to think about. Spawning a room and/or exit would be analog to *Gandalf saying "Mellon" at the west-gate*. Perhaps such could be modeled with objects and locks? (elvish-common dictionary, elvish speaker diploma and a lock on the west-gate).

2. Honoring the contents of a part or result is somehow analog to Aladdin finding the lamp with the Genie inside. However, such could be constructed as 2 different puzzles: one that takes `old dirty lamp` and produces `magic lamp` and a second that takes `magic lamp` and produces `empty magic lamp` and `genie`. Of course, it would require the playing user to invoke the `use` command with the *correct lamp* multiple times. 

@Griatch,  @pinkmagicdev,  @Zireael07, @Ockad, @samjaninf opinions?

Ockad commented:
Excellent job. Feels like a complete system now.
Thank you for your contributions!

Griatch commented:
@Henddher 

1. It seems this functionality would be resolved by allowing to optionally specify the typeclass of the component/result. Of course, you are then *really* close to making a Crafting system at the same time. But yes, normally things like access to exits are handled by locks or by assigning tags/attributes that pass locks. In code, the way to handle this would be to allow the recipe to accept a *callable* as a result. the callable gets fired with, say, the caller and the recipe when the recipe is complete - rather than spawn a result object that callable could then assign locks/attributes/tags to the caller as desired. For example, it could assign the "know_mellon" tag on the caller which is checked by the lock on the gate. 
2. Not sure how one would handle this beyond making two puzzles. If you make sure to remove the `old dirty lamp` and replace it with `magical lamp` etc (keeping the alias `lamp` in all cases), I suppose it work ok.

chainsol commented:
Not to butt in here, but I just wanted to mention I love the idea of this contrib! If it happens to be generic enough I'll be using it in Ainneve, especially if it uses tagged objects instead of a custom attribute!

Henddher commented:
Thank you @Griatch !

> 2. Not sure how one would handle this beyond making two puzzles. If you make sure to remove the `old dirty lamp` and replace it with `magical lamp` etc (keeping the alias `lamp` in all cases), I suppose it work ok.

Precisely! 

There are few interesting points about puzzles and the concept of puzzle-part-protptypes:

1. Remember that a puzzle recipe takes both parts and results and as I create each puzzle-part-prototype [using `proto_def` function](https://github.com/evennia/evennia/pull/1653/files#diff-0abe3dde8b11db34274d3377bb0c1b88R96), `typeclass`, `locks`, `tags`, `permissions` and other properties are all *remembered* within the recipe. Later, when the part and/or result is spawned, it's nearly a clone of the original object. 

2. By design, when a puzzle is resolved, the combined parts are destroyed automatically just before all results are spawned. An interesting aspect of this behavior is that a builder can use the same object any number of times, both as a part and as a result, within the same puzzle. 

My canonical example is the `makefire` puzzle: 
    stone + flint => fire, stone, and flint

**Builder:**
```
@create/drop fire
@create/drop stone
@create/drop flint

@puzzle makefire, stone, flint = fire, stone, flint

@armpuzzle #x
```

**Player:**
```
use stone, flint
... Fire appears
use stone, flint
... Fire appears
```

When the `makefire` puzzle is resolved, although the `use`d *stone* and *flint* have been destroyed, new ones are spawned along with a new *fire*. 
Consequently, a new `makefire` puzzle is left automatically `@arm`ed because the new *stone* and *flint* are `makefire` parts ready to be `use`d again. I.e., the puzzle becomes self-sustaining (infinite).

In my test cases, I underlined couple of scenarios to illustrate these *slick* capabilities:

- [chopping tree to get wood](https://github.com/evennia/evennia/pull/1653/files#diff-d530d174c30ebdcffafaf241fc3eec6dR2195)
- [digging a hole](https://github.com/evennia/evennia/pull/1653/files#diff-d530d174c30ebdcffafaf241fc3eec6dR2241)

In the Aladdin scenario, using these capabilities, I could even force the Genie back into the lamp:

```
@create/drop dirty lamp
@create/drop magical lamp
@create/drop Genie
@create/drop empty magical lamp

@puzzle maglamp, dirty lamp = magical lamp
@puzzle maglampgen, magical lamp = Genie, empty magical lamp
@puzzle maglamp2, empty magical lamp, Genie = magical lamp
```

**Player**
```
use dirty lamp
# dirty lamp is destroyed
... magical lamp appears

use magical lamp
# magical lamp is destroyed
... empty magical lamp appears
... Genie appears

use empty magical lamp, Genie
# empty magical lamp is destroyed
# Genie is destroyed
... magical lamp appears

```

Now, for the first point:

> 1. It seems this functionality would be resolved by allowing to optionally specify the typeclass of the component/result. Of course, you are then _really_ close to making a Crafting system at the same time. But yes, normally things like access to exits are handled by locks or by assigning tags/attributes that pass locks. In code, the way to handle this would be to allow the recipe to accept a _callable_ as a result. the callable gets fired with, say, the caller and the recipe when the recipe is complete - rather than spawn a result object that callable could then assign locks/attributes/tags to the caller as desired. For example, it could assign the "know_mellon" tag on the caller which is checked by the lock on the gate.

Indeed, when I create the puzzle-part-prototype [using `proto_def` function](https://github.com/evennia/evennia/pull/1653/files#diff-0abe3dde8b11db34274d3377bb0c1b88R96),  `typeclass` of the object is also *remembered* 
... I guess I should try building a puzzle with other type of objects and see what happens?!

I thought about allowing the Builder to "script" some of the behavior (using python code), that felt like a can of worms I didn't want to deal with during the first iteration of puzzles. It's totally worth considering for a future iteration :)

However, I was thinking more along the lines of *spawning* the actual exit, or the room, or the NPC. 

When it comes to exits and rooms, I now think it's an overkill because each time the puzzle got resolved, the new room and/or exit would be created. That would probably didn't work as there would be multiple exits with the same name at the same location. Maybe? Besides, as you confirmed, locks and permissions attached to exits and objects would do the trick just fine.

I wonder if the same trick could be used on NPCs?

... if so, I would not need to worry about neither of those and mark that TODO checkbox as completed :)

Griatch commented:
@Henddher I've not looked at the code in a while, but it seems you in the end went full out using actual objects to build your recipes rather than describing your recipes through tagging, is that correct?

Henddher commented:
Hi @Griatch!

> but it seems you in the end went full out using actual objects to build your recipes rather than describing your recipes through tagging, is that correct?

#### Yes and no ...

Builder creates objects but the whole puzzle system (arming and resolution) is all based on matching tags and spawned objects from prototypes and not from the original built objects.

#### Let me explain step by step:

1. Builder creates objects, sets typeclass, descriptions, locks, permissions, home, location, attributes, tags,  etc. etc. just like any other common object:

```
@create/drop blade
@describe blade = A kind and hard edge with a redish shine under sun light but cold shine under moon light.

@create/drop hilt
@describe hilt = A grip that easily accommodates both hands, a pommel holding a strange stone, and a two prone cross-guard.

@create/drop Andúril Flame of the West
@describe Andúril Flame of the West = Very bright was that sword when it was made whole again; the light of the sun shone redly in it, and the light of the moon shone cold, and its edge was hard and keen. And Aragorn gave it a new name and called it Andúril, Flame of the West. —J.R.R. Tolkien, The Lord of the Rings, The Fellowship of the Ring, Book Two, Chapter III: "The Ring goes South"
```

2. Then, builder invokes the `@puzzle` command specifying *puzzle name*, `parts` and `results`:

```
@puzzle anduril, hilt, blade = Andúril Flame of the West
```

3. Puzzles system derives `prototype`s from each specified `part` and `result` and adds 2 special puzzle tags to each prototype. One for the name of the puzzle `("anduril", "puzzles")` and the other for internal search/grouping `("puzzle_member", "puzzles")`. These prototypes are Evennia 0.8.0 compatible. This is done by the [using `proto_def` function](https://github.com/evennia/evennia/pull/1653/files#diff-0abe3dde8b11db34274d3377bb0c1b88R96)

```python
parts = [
{
    'typeclass': 'typeclasses.objects.Object', 
    'key':'hilt', 'description':'A grip th...', 
    'tags': [('anduril', 'puzzles'), ('puzzle_member', 'puzzles'), ... ],
    'prototype_key': '...',
     ...
},
{
    'typeclass': 'typeclasses.objects.Object', 
    'key':'blade', 'description': 'A kind and hard...', 
    'tags': [('anduril', 'puzzles'), ('puzzle_member', 'puzzles'), ... ],
    'prototype_key': '...',
    ...
},
 ]
results = [
{
    'typeclass': 'typeclasses.objects.Object', 
    'key':'Andúril Flame of the West', 'description': 'Very bright was that ...', 
    'tags': [('anduril', 'puzzles'), ('puzzle_member', 'puzzles'), ... ],
    'prototype_key': '...',
    ...
}
]
```

4. Then, a new `PuzzleRecipe` (subclass of `DefaultScript`) is created and saved containing `list` of `prototypes` in its `dict` representation of all parts and results and special tag `("puzzle_recipe", "puzzles")` for internal grouping/search.

```python
recipe = PuzzleRecipe()
recipe.db.puzzle_name = 'anduril'
recipe.db.parts = parts
recipe.db.results = results
recipe.tags.add(("puzzle_recipe", "puzzles"))
...
```

5. At this point, Builder should `@destroy` all objects: `hilt`, `blade`, and `Andúril Flame of the West`. They are NO longer needed for anything else. It's really up to the builder.

6. Builder `@arms` the **anduril** puzzle, using the recipe's `#dbref` (e.g. `#123`) and the puzzle system retrieves the puzzle recipe from DB, introspects all stored prototypes, and `spawn`s new objects for each `part` (None of the `result`s are spawned at this time).

```
@arm #123
...
hilt #3 created at ...
blade #4 created at ...
```

7. Builder is now done.

.... At a later time ...

8. A Player finds `hilt` and `blade` and issues the `use` command:

```
use hilt, blade
```

9. Puzzle system searches for all puzzle recipes that have parts `hilt` and `blade`. It finds the *anduril* puzzle and proceeds to `destroy` the passed in `hilt` and `blade` and `spawn`s a new `Andúril Flame of the West` object.

```
Andúril Flame of the West appears
```

10. Player is now done and he/she can take and use `Andúril Flame of the West` as it's just a regular world object.

#### Does this explain the process?

Henddher commented:
Let me explain the actual `part`s matching process that occurs when Player issues the `use` command:

Crete lookup `dict`s: 

```python
parts_dict # key: part.dbref; value: part
puzzlename_tags_dict # key: puzzle name; value = list of all parts that contain tag (puzzle_name, "puzzles")
puzzle_ingredients # key: part.dbref; value: part prototype
```
https://github.com/evennia/evennia/pull/1653/files#diff-0abe3dde8b11db34274d3377bb0c1b88R545

Find all puzzle recipes matching all puzzle names of all passed in parts:
https://github.com/evennia/evennia/pull/1653/files#diff-0abe3dde8b11db34274d3377bb0c1b88R563

For all puzzle recipes retrieved above (by puzzle name), check if all its parts were passed in. The check is done by comparing the prototype (from `puzzle_ingredients`) of each part against each of the prototypes stored within `puzzle.db.parts`.  If so, a new puzzle that can be resolved is added to `matched_puzzles`.
https://github.com/evennia/evennia/pull/1653/files#diff-0abe3dde8b11db34274d3377bb0c1b88R580

Henddher commented on evennia/prototypes/spawner.py:
```diff
@@ -716,7 +716,7 @@ def spawn(*prototypes, **kwargs):
         tags = []
         for (tag, category, data) in val:
-            tags.append((init_spawn_value(val, str), category, data))
+            tags.append((init_spawn_value(tag, str), category, data))
```
Hmmm ... 
[comment](https://github.com/evennia/evennia/pull/1653#issuecomment-427663883) 

Didn't you pick this change on your spawner fix @Griatch?

Griatch replied:
I did, yes. Not sure why it's still there. Fixed (again) in master now :)

Henddher commented:
Hi @Griatch 

There seems to be something wrong with `evennia.utils.search.search_script_attribute`:

```python
            _puzzles = search.search_script_attribute(
                    key='puzzle_name',
                    value=puzzle_name              
            )
```

It's always returning empty `<QuerySet>` even though the `script` with the `puzzle_name` attribute is there.

To reproduce the issue, any vanilla puzzle will suffice. Of course, creating a `script` and adding an `attribute` and then searching by that `attribute` should demonstrate the problem too.

I am testing puzzles on my docker image and obviously there is something else wrong as all tests are passing but real puzzle system isn't working.

Griatch commented:
@Henddher 
Thanks for the explanation. So the rewrite is *not* based on tags in the sense that I was envisioning it, but kept the old object-based functionality but is is using tags for identification under the hood. Fair enough. 

I updated master with that tag iteration fix. Weird, could have sworn I applied it already. 

As for the unicode vs string, Just creating a script, assigning an attribute and then searching for it works fine, as you've found out. So is the issue that something in your system assigns unicode or is there some issue with reading back strings from the Attribute? Trying to figure out of this is something I need to look at from the Evennia side or not.

Henddher commented:
> So the rewrite is _not_ based on tags in the sense that I was envisioning it, but kept the old object-based functionality but is is using tags for identification under the hood. Fair enough.

True. It's not exactly as you originally suggested in google group: a puzzle builder doesn't need to write a 'dict'-like description of the puzzle with tags:

```python
sword_recipe = ("blade", "hilt", "pommel")
```

Instead, the puzzle builder creates the objects, apply tags, properties, attributes, etc. (using vanilla *builder* commands) and the puzzle system creates `prototype`s.  Then, when the puzzle is armed and/or resolved, the `prototypes` are used to `spawn`ed new objects which are *clones* of the originals.

```python
proto_def = {
'locks': 'call:true();control:id(1) or perm(Admin);delete:id(1) or perm(Admin);edit:perm(Admin);examine:perm(Builder);get:all();puppet:pperm(Developer);tell:perm(Admin);view:all()', 
'location': <Room: Limbo>, 
'key': u'stone', 
'home': <Room: Limbo>, 
'typeclass': u'typeclasses.objects.Object', 
'permissions': [], 
'desc': 'You see nothing special.'
}

```

Although I am biased, I find puzzle recipe with prototypes easier to maintain because a builder just needs to create regular objects with all its attributes once, create a `@puzzle` recipe once, and `@armpuzzle` the parts as many times as needed. From that point on, depending on the nature of the puzzle (self-sustaining [1], multiplicative [2], bidirectional [3], or what have you), puzzle parts and results are spawned automatically with no Builder intervention. I'm biased though 😈 😃.

I do think that if I tackle the Crafting system, I'll go with a pure-tags approach as you originally suggested. The craft system doesn't need to be as rigorous as the puzzle system is.

(Did I just volunteer to implement the Crafting system?! :stuck_out_tongue_winking_eye:)

> As for the unicode vs string, Just creating a script, assigning an attribute and then searching for it works fine, as you've found out. So is the issue that something in your system assigns unicode or is there some issue with reading back strings from the Attribute? Trying to figure out of this is something I need to look at from the Evennia side or not. 

I didn't use any `unicode` to `str` conversion (or vice versa) anywhere in the puzzles module but I will compare with other modules to see if they use special functions that do conversions. I might have forgotten to use them.

I didn't see any `u'...'` output during my development in 0.7.0 but now in 0.8.0 I am noticing them everywhere. I'll let you know what I find out, @Griatch.

Henddher commented:
@Griatch ... something's off with `str` and `unicode`.

Here is a simple experiment:

I create a regular object (`@create/drop stone`).

Here is the output:

![screen shot 2018-10-13 at 1 59 31 pm](https://user-images.githubusercontent.com/3991010/46908959-61052480-cef0-11e8-81b5-02bac0b63d39.png)

The `stone` object's `desc` is a `str`, however its `key` is `unicode`. `key` came from the web-interface while `desc` was automatically produced by the command set.

Notice what happens when I set the `desc` (`@describe stone = massive boulder`)

![screen shot 2018-10-13 at 2 04 29 pm](https://user-images.githubusercontent.com/3991010/46908993-f2749680-cef0-11e8-8694-0d9137af8c9a.png)

**Now the `stone`s description `desc` is a `unicode` string**

The *description* entered is now a `unicode` string. 

Either it was saved as `unicode` or retrieved from DB as `unicode`. 

**Doesn't this look like a systemic issue? (instead of a puzzles only issue) @Griatch ?**

Griatch commented:
@Henddher 
Well, usually it doesn't really matter - in Python2, you commonly use unicode and bytestrings interchangeably. You seem to have found a case where it does matter and sending the raw input will lead to a parsing problem.

Henddher commented:
> You seem to have found a case where it does matter and sending the raw input will lead to a parsing problem.
> In light of the upcoming change to python 3, I don't think we should go into making any big unicode/bytestring changes at this point

Right - no point in reinventing the wheel when python 3 handles all strings as unicode always. 

I was just pointing out what it seemed like a systemic problem to me. I was suspecting - still am - that it probably was introduced in 0.8.0 knowing that didn't exist in 0.7.0.

> making sure to convert the search term to a string will resolve this issue.

Fair enough 👍  it's probably handful of places within the puzzles module. I'll work around it 👍

Henddher commented:
I'm nearly much done with this contrib ... just a few touches left.

I tested using `DefaultChar` as part and result in a puzzle and it worked (of course, with local changes that relax proto_def storing and matching).

I will add a new attribute to puzzle recipe that will allow Builder to specify which proto keys are to be "stored" and "matched" during puzzle resolution. Something like `"masks" : ["+key", "-location", "+desc", "-typeclass"]`. 

For instance, specifying `"masks": ["-typeclass"]` would exclude "typeclass" during the matching parts phase of the `use` command. I think `+<a key>` will be the default and `-<a key>` the overriding value to exclude `a-key` during matching.

Consequently, allowing the exclusion of many of the keys would essentially allow the fulfillment of the original suggestion for a craft system.

Henddher commented:
### outstanding issues:

- [ ] uppercase letter in puzzle name fails to be resolved.
5ac40e0 vs 59e29cf

- [ ] results cannot be empty but it's a good use case for a *red herring*. See 5ac40e0

Griatch commented:
@Henddher This appears to still be WIP? It's probably time to wrap this long-standing PR up now. :)

Henddher commented:
Hello @Griatch 

I just spoke with my *Product Owner* (my 15yo son 😛) about all left over tasks and limitations and he signed off on them. IOW, he's satisfied with how the system works overall. 

When we discussed the issue of not being able to honor the contents of a result and/or part of a puzzle (because of how complicated `spawn`ing would need to be), and the simplicity of the workaround of creating *many-objects-that-look-like-the-same-object* ([Aladdin's Genie and Magic Lamp](https://github.com/evennia/evennia/pull/1653#issuecomment-427673140)) he was satisfied. He was also thinking he could use a Lock with visibility *none* for the case of a chest with treasure (chest's description showing it has treasure and treasure being on the room as invisible). IOW, another way to solve the same problem using Locks.

In summary, **this is ready for review**

Henddher commented:
@Griatch How should I go about the Wiki? and documenting how each command is meant to be used and examples?

I think the Contrib Wiki is not auto-generated from source code as other wiki pages are? (or am I wrong?)

Henddher commented on evennia/contrib/tests.py:
```diff
@@ -1717,6 +1717,710 @@ def test_generate(self):
+        parts = ['Balloon']
+        results = ['Balloon']
+        recipe_dbref = self._good_recipe(
+            'boom!!!',  # FIXME: uppercase name fails
```
This is still outstanding. 
Uppercase letters cause failures. Needs investigation but seems like outside puzzles modules and inherent to search.

Henddher replied:
Thanks @Griatch!

I removed the `FIXME:` as it's case-insensitive is the expected behavior.

Henddher commented:
Hi @Griatch!
and Happy New Year 👍 

As I mentioned [earlier](https://github.com/evennia/evennia/pull/1653#issuecomment-443525532), I think this contribution is ready for code review. it's got nice coverage []() too. 

You think there is anything I should improve?

Please let me know. I've got couple of ideas I'd like to work on right after this is mainlined.

Henddher commented:
Great!!! Thank you @Griatch!

Next month, I will start the discussion surrounding Craft system if that's OK with you.
